parsers/ES.py

```diff
@@ -170,69 +171,55 @@ def fetch_production(
-
-    if island_data:
-        for response in island_data:
-            if response.production() >= 0:
-                response_data = {
-                    "datetime": get(response.timestamp).datetime,
-                    "production": {
-                        "biomass": response.waste,
-                        "coal": response.carbon,
-                        "gas": round((response.gas + response.combined), 2),
-                        "geothermal": 0.0,
-                        "hydro": response.hydraulic,
-                        "nuclear": 0.0,
-                        "oil": round((response.vapor + response.diesel), 2),
-                        "solar": response.solar,
-                        "unknown": response.other,
-                        "wind": response.wind,
-                    },
-                    "source": "demanda.ree.es",
-                    "storage": {"hydro": 0.0, "battery": 0.0},
-                    "zoneKey": zone_key,
-                }
-
-                ### Zone overrides
+    for event in island_data:
+        production = ProductionMix()
+        storage = StorageMix()
+        for mode in [
+            attribute
+            for attribute in getmembers(event)
+            if not attribute[0].startswith("__")
+        ]:
+            if mode[0] in PRODUCTION_MAPPING.keys():
+                ## Production mapping override for Canary Islands
                 # NOTE the LNG terminals are not built yet, so power generated by "gas" or "combined" in ES-CN domain is actually using oil.
                 # Recheck this every 6 months and move to gas key if there has been a change.
                 # Last checked: 2022-06-27
                 if zone_key.split("-")[1] == "CN":
-                    response_data["production"]["gas"] = 0.0
-                    response_data["production"]["oil"] = round(
-                        (
-                            response.vapor
-                            + response.diesel
-                            + response.gas
-                            + response.combined
-                        ),
-                        2,
+                    PRODUCTION_MAPPING.update(
+                        {
+                            "gas": "oil",
+                            "combined": "oil",
+                        }
                     )
+                    # Reset gas to 0, as it is now oil, normally this should be skipped or set to None but to conform to how REE does it, we set it to 0
+                    # We should aim to change this upstream in the future to make it uniform with other parsers going forward.
+                    production.gas = 0
 
-                # Hydro response is hydro storage
-                if zone_key == "ES-CN-HI":
-                    response_data["production"]["hydro"] = 0.0
-                    response_data["storage"]["hydro"] = -response.hydraulic
+                production.add_value(mode=PRODUCTION_MAPPING[mode[0]], value=mode[1])
```

Comment:
How about maybe adding some logic inside add_value so that if the production mode is None, it sets it to zero and then add the value ?
I think it would be a bit more safe ? And would avoid some of the code above.